Feature/list map

ColmBhandal opened the pull request:
## Description

Refactored the MultiValueMap to something more abstract. Renamed the types to be less ambiguous. Added a list-valued map.

As an aside, removed the getHashCode() function and replaced with equals with a `DictEquals` function.

ColmBhandal committed:
Replaced MultiValueMap with SetDictionary and abstracted collection-based functionality out

ColmBhandal committed:
Renamed types to be less ambiguous

C# already has a library with a ListDictionary in it which means something else
So to disambiguate, the names of these collection-valued dicts include the term "Valued" indicating that it's the values at each key that is a collection

ColmBhandal committed:
List valued dictionary plus tests

ColmBhandal committed:
Added a null check to the dictionary

The C# underlying dictionary will throw this anyway, but no harm pushing the error up a level in the stack trace

ColmBhandal committed:
InsertAtIndex function and tests

This function allows a user to insert an element to a given index in the list at the given key
One test case is where there is already a non-mepty list mapped from the given key
Another test case is where there is no list mapped to the given key- confirming that insert creates one

ColmBhandal committed:
Updated changelog

PR number on each line relating to this PR
Added a line about list-valued dictionaries

hctrdev commented on CsharpExtras/Api/CsharpExtrasApi.cs:
```diff
@@ -22,14 +23,19 @@ public void SetLogger(ILogger logger)
             => new RegexPatternDictionaryImpl<TVal>();
-        public IMultiValueMap<TKey, TVal> NewMultiValueMap<TKey, TVal>()
-            => new MultiValueMapImpl<TKey, TVal>();
+        public ISetValuedDictionary<TKey, TVal> NewSetDictionary<TKey, TVal>()
```
I would rename this to `NewSetValuedDictionary` to match the type and the naming of the method to get a new list valued dictionary.

hctrdev commented on CsharpExtras/Api/ICsharpExtrasApi.cs:
```diff
@@ -26,7 +27,7 @@ public interface ICsharpExtrasApi
         ILazyFunctionMap<TKey, TVal> NewLazyFunctionMap<TKey, TVal>(Func<TKey, TVal> backingFunction);
         ILeafBase<TVal> NewLeaf<TVal>(TVal payload);
-        IMultiValueMap<TKey, TVal> NewMultiValueMap<TKey, TVal>();
+        ISetValuedDictionary<TKey, TVal> NewSetDictionary<TKey, TVal>();
```
Same as above.

hctrdev commented on CsharpExtras/Dictionary/Collection/ISetValuedDictionary.cs:
```diff
@@ -0,0 +1,12 @@
+    public interface ISetValuedDictionary<TKey, TVal>
+        : ICollectionValuedDictionary<TKey, TVal, ISet<TVal>>
+    {
+        ISetValuedDictionary<TKey, TOther> TransformValues<TOther>(Func<TVal, TOther> transformer);
```
Shouldn't this shortcut also exist on the `IListValuedDictionary`?

hctrdev commented on CsharpExtrasTest/Dictionary/Collection/ListValuedDictionaryTest.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        [Test]
+        public void Given_ListDictionaries_When_SameElementsAddedInSameOrderPerKey_Then_DictionariesEqual()
```
Did you add a test for the reverse of this?
If the same elements are added in a different order per key, then DictEquals will return false.

hctrdev commented on CsharpExtrasTest/Dictionary/Collection/SetDictionaryTest.cs:
```diff
@@ -4,19 +4,36 @@
 {
     [TestFixture]
-    class MultiValueMapTest
+    class SetDictionaryTest
```
Rename to `SetValuedDictionaryTest` to be consistent.

ColmBhandal committed:
PR fixes

Renamed API function
Added TransformValues to list valued dictionary
Added list valued dictionary test case

Signed-off-by: cbhandal <bhandalc@gmail.com>

ColmBhandal committed:
Renamed test file & class to be consistent

ColmBhandal committed:
Fixed YAML to use "develop" instead of "master"
